reconcile/terraform_resources_cloudflare.py

```diff
@@ -0,0 +1,283 @@
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.terraform import safe_resource_id
+from reconcile.utils.terraform.config_client import TerraformConfigClientCollection
+from reconcile.utils.terraform_client import TerraformClient
+from reconcile.utils.terrascript.cloudflare_client import (
+    CloudflareAccountConfig,
+    TerraformS3BackendConfig,
+    TerrascriptCloudflareClient,
+    create_cloudflare_terrascript,
+)
+from reconcile.utils.helpers import filter_null
+
+QONTRACT_INTEGRATION = "terraform_resources_cloudflare"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+QONTRACT_TF_PREFIX = "qrtfcf"
+
+
+def create_cloudflare_account_config(
+    settings: dict[str, Any], cf_acct: CloudflareAccountV1
+) -> CloudflareAccountConfig:
+    secret_reader = SecretReader(settings=settings)
+    cf_acct_creds = secret_reader.read_all({"path": cf_acct.api_credentials.path})
+    return CloudflareAccountConfig(
+        cf_acct.name,
+        cf_acct_creds.get("api_token"),
+    )
+
+
+def create_backend_config(
+    settings: dict[str, Any],
+    aws_acct: AWSAccountV1,
+    cf_acct: CloudflareAccountV1,
+) -> TerraformS3BackendConfig:
+    secret_reader = SecretReader(settings=settings)
+    aws_acct_creds = secret_reader.read_all({"path": aws_acct.automation_token.path})
+
+    # default from AWS account file
+    tf_state = aws_acct.terraform_state
+    if tf_state is None:
+        raise ValueError(
+            f"AWS account {aws_acct.name} cannot be used for Cloudflare "
+            f"account {cf_acct.name} because it does define a terraform state "
+        )
+
+    integrations = tf_state.integrations or []
+    for i in integrations or []:
+        name = i.integration
+        if name.replace("-", "_") == QONTRACT_INTEGRATION:
+            # we have to ensure the bucket key(file) is unique across
+            # Cloudflare accounts to support running per-account
+            bucket_key = f"{QONTRACT_INTEGRATION}-{cf_acct.name}.tfstate"
+            bucket_name = tf_state.bucket
+            bucket_region = tf_state.region
+            break
+
+    if bucket_name and bucket_key and bucket_region:
+        backend_config = TerraformS3BackendConfig(
+            aws_acct_creds["aws_access_key_id"],
+            aws_acct_creds["aws_secret_access_key"],
+            bucket_name,
+            bucket_key,
+            bucket_region,
+        )
+    else:
+        raise ValueError(f"No state bucket config found for account {aws_acct.name}")
+
+    return backend_config
+
+
+def get_resources(
+    query_data: TerraformResourcesCloudflareQueryData,
+) -> list[NamespaceTerraformProviderResourceCloudflareV1]:
+    """Get all Cloudflare V1 resources from the Cloudflare query data"""
+    return [
+        res
+        for namespace in filter_null(query_data.namespaces)
+        for res in filter_null(namespace.external_resources)
+        if isinstance(res, NamespaceTerraformProviderResourceCloudflareV1)
```

Comment:
In addition to what Steve said, we could add `isSearchable` attribute here https://github.com/app-sre/qontract-schemas/blob/main/graphql-schemas/schema.yml#L1114 . This in theory should help us remove this check and do filtering on server side. The change should also be backwards compatible :crossed_fingers:

Reply:
If it becomes `V2` then we have to update the integration, isn't that the point of versions?

I don't think `isSearchable` helps because that enables filtering based on fields & values, but this is about detecting the type here.

I don't like this check more than anyone else but I could not find a way to avoid it. This is needed because when qenerate generates classes and it find a field set to an interface it generates a `Union` type with the interface class and whatever types were requested in the query, in this case `NamespaceTerraformProviderResourceCloudflareV1`

Example in `vpc_peerings_validator`:
* The [connections field is an interface](https://github.com/app-sre/qontract-schemas/blob/b41ebc6c51c4c75ece2ebe58eddd642589c2a5c8/graphql-schemas/schema.yml#L688-L704)
* The graphql query [selects only the two relevant types](https://github.com/app-sre/qontract-reconcile/blob/d56f51c03ec574f4b380a8d6597ad4552ea799bf/reconcile/gql_definitions/vpc_peerings_validator/vpc_peerings_validator.gql#L13-L30)
* The [Union type field](https://github.com/app-sre/qontract-reconcile/blob/d56f51c03ec574f4b380a8d6597ad4552ea799bf/reconcile/gql_definitions/vpc_peerings_validator/vpc_peerings_validator.py#L132-L139) generated only has the two selected types, in addition to the interface

Reply:
The code might need to change with a `V2`, but the query to get Cloudflare resources may not necessarily need to. My thought was just that if `provider` on `NamespaceTerraformProviderResourceCloudflareV1` is `cloudflare` then this could be avoided?  With that being said, I get that then we can't provide the actual return type. We can revisit if we end up needing a `V2`.

Reply:
I thought about checking the `provider` field but even though that field is set on the [NamespaceExternalResource_v1 interface](https://github.com/app-sre/qontract-schemas/blob/main/graphql-schemas/schema.yml#L1118), qenerate does not add it to the `NamespaceExternalResourceV1` class and so mypy complains with `"NamespaceExternalResourceV1" has no attribute "provider"  [attr-defined]`